File: src/utils/ReleaseLoader.ts
import { Release, ReleaseVersion, ReleaseSource } from "../types/Release";
import semver from "semver";
import { v4 as uuidv4 } from "uuid";

let releases: Release[] = [];
let sources: ReleaseSource[] = [];
let releaseSourcesUrl: string = 'https://raw.githubusercontent.com/jxnxsdev/TidaLuna-Installer/main/resources/sources.json';
let releasesLoaded: boolean = false;

/**
 * Loads releases from sources whuch supply a direct download link to a json file
 * @param source - The source to load releases from
 * @returns - An array of releases
 */
async function processDirectReleaseSource(source: ReleaseSource): Promise<Release[]> {
    const response = await fetch(source.url);
    if (!response.ok) {
        console.error(`Failed to fetch release data from ${source.url}: ${response.statusText}`);
    }
    const data = await response.json();
    for (const release of data.releases) {
        release.id = await uuidv4();
    }
    return data.releases;
}


/**
 * Extracts the release channel name from a tag.
 * For example: "1.2.3-alpha", "alpha-1.2.3", "dev" → "alpha", "alpha", "dev"
 */
function extractChannelName(tag: string): string {
    const clean = tag.replace(/^v/, '');

    // Try to parse as semver and get prerelease label
    const parsed = semver.parse(clean);
    if (parsed && parsed.prerelease.length > 0) {
        const first = String(parsed.prerelease[0]);
        return first.split(/[-.]/)[0]; // e.g., "alpha-hotfix" → "alpha"
    }

    // Try matching "alpha-1.2.3" format
    const match = clean.match(/^([a-zA-Z]+)[-_]\d/);
    if (match) return match[1];

    // Valid semver with no prerelease → stable
    if (semver.valid(clean)) return "stable";

    // Fallback: whole tag
    return clean;
}


/**
 * Loads releases from GitHub and groups them into release channels
 * @param source - The source to load releases from
 * @returns - An array of releases
 */
export async function processGithubReleaseSource(source: ReleaseSource): Promise<Release[]> {
    const releaseUrl = `https://api.github.com/repos/${source.url}/releases`;
    const response = await fetch(releaseUrl);
    if (!response.ok) {
        console.error(`Failed to fetch release data from ${releaseUrl}: ${response.statusText}`);
        return [];
    }

    const data = await response.json();

    const grouped: Map<string, Release> = new Map();

    for (const release of data) {
        const tag = release.tag_name;
        const channelName = extractChannelName(tag);

        const version: ReleaseVersion = {
            version: tag,
            download: `https://github.com/${source.url}/releases/download/${tag}/luna.zip`
        };

        if (!grouped.has(channelName)) {
            grouped.set(channelName, {
                name: channelName,
                githubUrl: `https://github.com/${source.url}`,
                id: uuidv4(),
                versions: [version]
            });
        } else {
            grouped.get(channelName)!.versions.push(version);
        }
    }

    return Array.from(grouped.values());
}


/**
 * Loads releases from the sources defined in the sources.json file
 * 
 */
async function loadReleaseSources() {
    const response = await fetch(releaseSourcesUrl);
    if (!response.ok) {
        throw new Error(`Failed to fetch release sources: ${response.statusText}`);
        return;
    }
    sources = await response.json();
}

/**
 * Loads releases from the sources defined in the sources.json file
 * @returns - An array of releases
 */
export async function loadReleases() {
    if (releasesLoaded) return releases;
    await loadReleaseSources();
    for (const source of sources) {
        if (source.type === 'github') {
            const githubReleases = await processGithubReleaseSource(source);
            releases.push(...githubReleases);
        } else if (source.type === 'direct') {
            const directReleases = await processDirectReleaseSource(source);
            releases.push(...directReleases);
        }
    }
    releasesLoaded = true;
    return releases;
}

File: src/utils/release_loader.rs
use crate::types::types::{Release, ReleaseVersion, ReleaseSource, ReleaseSourceType};
use reqwest::Client;
use semver::Version;
use uuid::Uuid;
use std::collections::HashMap;

pub struct ReleaseLoader {
    pub releases: Vec<Release>,
    pub sources: Vec<ReleaseSource>,
    pub release_sources_url: String,
    pub releases_loaded: bool,
    client: Client,
}

impl ReleaseLoader {
    pub fn new(release_sources_url: &str) -> Self {
        Self {
            releases: Vec::new(),
            sources: Vec::new(),
            release_sources_url: release_sources_url.to_string(),
            releases_loaded: false,
            client: Client::new(),
        }
    }

    pub async fn load_release_sources(&mut self) -> anyhow::Result<()> {
        let resp = self.client.get(&self.release_sources_url).send().await?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch release sources: {}", resp.status());
        }
        self.sources = resp.json().await?;
        Ok(())
    }

    async fn process_direct_release_source(&self, source: &ReleaseSource) -> anyhow::Result<Vec<Release>> {
        let resp = self.client.get(&source.url).send().await?;
        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch direct release: {}", resp.status());
        }
        let mut releases: Vec<Release> = resp.json().await?;
        for release in &mut releases {
            release.id = Uuid::new_v4().to_string();
        }
        Ok(releases)
    }

    fn extract_channel_name(tag: &str) -> String {
        let clean = tag.trim_start_matches('v');

        if let Ok(parsed) = Version::parse(clean) {
            if !parsed.pre.is_empty() {
                let ver_str = parsed.to_string();
                let prerelease = ver_str.split('-').nth(1).unwrap_or("unknown");
                return prerelease.split('.').next().unwrap_or("unknown").to_string();
            }
            return "stable".to_string();
        }

        if let Some(caps) = regex::Regex::new(r"^([a-zA-Z]+)[-_]\d").unwrap().captures(clean) {
            return caps.get(1).unwrap().as_str().to_string();
        }

        clean.to_string()
    }

    async fn process_github_release_source(&self, source: &ReleaseSource) -> anyhow::Result<Vec<Release>> {
        let release_url = format!("https://api.github.com/repos/{}/releases", source.url);
        let resp = self.client
            .get(&release_url)
            .header("User-Agent", "tidaluna-installer")
            .send()
            .await?;

        if !resp.status().is_success() {
            anyhow::bail!("Failed to fetch GitHub release: {}", resp.status());
        }

        let data: serde_json::Value = resp.json().await?;
        let mut grouped: HashMap<String, Release> = HashMap::new();

        if let Some(array) = data.as_array() {
            for release in array {
                if let Some(tag) = release.get("tag_name").and_then(|v| v.as_str()) {
                    let channel_name = Self::extract_channel_name(tag);
                    let version = ReleaseVersion {
                        version: tag.to_string(),
                        download: format!("https://github.com/{}/releases/download/{}/luna.zip", source.url, tag),
                    };

                    grouped.entry(channel_name.clone())
                        .and_modify(|r| r.versions.push(version.clone()))
                        .or_insert(Release {
                            id: Uuid::new_v4().to_string(),
                            name: channel_name,
                            github_url: Some(format!("https://github.com/{}", source.url)),
                            versions: vec![version],
                        });
                }
            }
        }

        Ok(grouped.into_values().collect())
    }

    pub async fn load_releases(&mut self) -> anyhow::Result<&Vec<Release>> {
        if self.releases_loaded {
            return Ok(&self.releases);
        }

        self.load_release_sources().await?;

        for source in &self.sources {
            match source.source_type {
                ReleaseSourceType::Github => {
                    let github_releases = self.process_github_release_source(source).await?;
                    self.releases.extend(github_releases);
                }
                ReleaseSourceType::Direct => {
                    let direct_releases = self.process_direct_release_source(source).await?;
                    self.releases.extend(direct_releases);
                }
            }
        }

        self.releases_loaded = true;
        Ok(&self.releases)
    }
}
